samplekotlin/src/main/java/com/example/chattutorial/ChannelActivity3.kt

```diff
@@ -63,39 +62,36 @@ class ChannelActivity3 : AppCompatActivity(R.layout.activity_channel_3) {
             messageInputViewModel.editMessage.postValue(it)
         }
 
-        // custom typing info header bar
-        val channelController = ChatDomain.instance().useCases.watchChannel(cid, 0).execute().data()
-        val typingObserver = Observer<List<User>> { users ->
-            channelHeaderSub.text = if (users.isNotEmpty()) {
-                "typing: " + users.joinToString(", ") { it.extraData["name"] as String }
-            } else {
-                "nobody is typing"
-            }
-        }
-        channelController.typing.observe(this, typingObserver)
-
         // step 6 - handle back button behaviour correctly when you're in a thread
-        val backButtonHandler = {
+        channelHeaderView.onBackClick = {
             messageListViewModel.onEvent(MessageListViewModel.Event.BackButtonPressed)
         }
-        channelHeaderView.onBackClick = { backButtonHandler() }
 
-        onBackPressedDispatcher.addCallback(
-            this,
-            object : OnBackPressedCallback(true) {
-                override fun handleOnBackPressed() {
-                    backButtonHandler()
+        onBackPressedDispatcher.addCallback(this) {
+            channelHeaderView.onBackClick()
+        }
+
+        // custom typing info header bar
+        val nobodyTyping = "nobody is typing"
+
+        typingHeader.text = nobodyTyping
+
+        ChatDomain
+            .instance().useCases.watchChannel(cid, 0)
+            .execute()
+            .data().typing
```

Comment:
Might make this more readable if we extracted the chain until `data()` into an intermediate variable, to show that that returns a `ChannelController`, which can then be used for various channel-related actions. We currently go through a lot of types and instances of things hidden in this one long chain.

Reply:
True... what about a `.let { channelController -> }`? Static linting will probably highlight it as unneeded, though... Happy to go either way.

Reply:
I guess either of them achieves the same, though going without `let` seems simpler to me. Whichever you prefer 😄